src/XrdPfc/XrdPfc.cc

```diff
@@ -899,6 +907,78 @@ int Cache::LocalFilePath(const char *curl, char *buff, int blen,
    return -ENOENT;
 }
 
+//______________________________________________________________________________
+// If supported, write file_size as xattr to cinfo file.
+//------------------------------------------------------------------------------
+void Cache::WriteFileSizeXAttr(int cinfo_fd, long long file_size)
+{
+   if (m_metaXattr) {
+      int res = XrdSysXAttrActive->Set("pfc.fsize", &file_size, sizeof(long long), 0, cinfo_fd, 0);
+      if (res != 0) {
+         TRACE(Debug, "WriteFileSizeXAttr error setting xattr " << res);
+      }
+   }
+}
+
+//______________________________________________________________________________
+// Determine full size of the data file from the corresponding cinfo-file name.
+// Attempts to read xattr first and falls back to reading of the cinfo file.
+// Returns -error on failure.
+//------------------------------------------------------------------------------
+long long Cache::DetermineFullFileSize(const std::string &cinfo_fname)
+{
+   if (m_metaXattr) {
+      char pfn[4096];
+      m_oss->Lfn2Pfn(cinfo_fname.c_str(), pfn, 4096);
+      long long fsize = -1ll;
+      int res = XrdSysXAttrActive->Get("pfc.fsize", &fsize, sizeof(long long), pfn);
+      if (res == sizeof(long long))
+      {
+         return fsize;
+      }
+      else
+      {
+         TRACE(Debug, "DetermineFullFileSize error getting xattr " << res);
+      }
+   }
+
+   XrdOssDF *infoFile = m_oss->newFile(m_configuration.m_username.c_str());
+   XrdOucEnv env;
+   int res = infoFile->Open(cinfo_fname.c_str(), O_RDONLY, 0600, env);
+   if (res < 0)
+      return res;
+   Info info(m_trace, 0);
+   if ( ! info.Read(infoFile, cinfo_fname.c_str()))
+      return -EBADF;
+   return info.GetFileSize();
+}
+
+//______________________________________________________________________________
+// Calculate if the file is to be considered cached for the purposes of
+// only-if-cached and setting of atime of the Stat() calls.
+// Returns true if the file is to be conidered cached.
+//------------------------------------------------------------------------------
+bool Cache::DecideIfConsideredCached(long long file_size, long long bytes_on_disk)
+{
+   if (file_size == 0 || bytes_on_disk >= file_size)
+      return true;
+
+   double frac_on_disk = (double) bytes_on_disk / file_size;
+
+   if (file_size <= m_configuration.m_onlyIfCachedMinSize)
+   {
+      if (frac_on_disk >= m_configuration.m_onlyIfCachedMinFrac)
+         return true;
+   }
+   else
+   {
+      if (bytes_on_disk >= m_configuration.m_onlyIfCachedMinSize &&
+          frac_on_disk  >= m_configuration.m_onlyIfCachedMinFrac)
```

Comment:
I wonder if this should be `||` instead of `&&`, but it probably doesn't make much of a difference.

Reply:
Ok, great, thanks!